README.md
# Decrypt filter plugin for Embulk

Converts columns using an encryption algorithm such as AES.

Encrypted data is encoded using base64 or hex. For example, if you have following input records (password column was encrypted):

* Base64
```csv
id,password,comment
1,XaBAt/J3LNqKCVlWbu2E+g==,a
2,gUzzC+nJSBLbPTAzJlbbMA==,b
```
* Hex
```csv
id,password,comment
1,5DA040B7F2772CDA8A0959566EED84FA,a
2,814CF30BE9C94812DB3D30332656DB30,b
``` 
    
You can apply decryption to the password column and get following outputs:
```csv
id,password,comment
1,super,a
2,secret,b
``` 

## Overview

* **Plugin type**: filter

## Configuration

- **algorithm**: encryption algorithm (see below) (enum, required)
- **column_names**: names of string columns to encrypt (array of string, required)
- **key_type**: encryption key (enum, optional, default: inline), can be either "inline" or "s3"
- **key_hex**: encryption key (string, required if key_type is inline)
- **iv_hex**: encryption initialization vector (string, required if mode of the algorithm is CBC and key_type is inline)
- **input_encoding**: the encoding of encrypted value, can be either "base64" or "hex" (base16)
- **aws_params**: AWS/S3 parameters (hash, required if key_type is s3)
    - **region**: a valid AWS region
    - **access_key**: a valid AWS access key
    - **secret_key**: a valid AWS secret key
    - **bucket**: a valid S3 bucket
    - **path**: a valid S3 key (S3 file path)
    
S3 key file should be in valid YAML format: (iv_hex is required if mode of the algorithm is CBC)

```yaml
key_hex: 098F6BCD4621D373CADE4E832627B4F60A9172716AE6428409885B8B829CCB05
iv_hex: C9DD4BB33B827EB1FBA1B16A0074D460
```

## Algorithms

Available algorithms are:

* AES-256-CBC
* AES-192-CBC
* AES-128-CBC
* AES-256-ECB
* AES-192-ECB
* AES-128-ECB

## Example

* Inline key type

```yaml
 filters:
   - type: decrypt
     algorithm: AES-256-CBC
     column_names: [password, ip]
     key_hex: 098F6BCD4621D373CADE4E832627B4F60A9172716AE6428409885B8B829CCB05
     iv_hex: C9DD4BB33B827EB1FBA1B16A0074D460
     input_encoding: hex
 ```
* S3 key type

```yaml
 filters:
   - type: decrypt
     algorithm: AES-256-CBC
     input_encoding: hex
     column_names: [password, ip]
     key_type: s3
     aws_params:
       region: us-east-2
       access_key: XXXXXXXXXXXXXXXXXXXX
       secret_key: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
       bucket: com.sample.keys
       path: key.aes
```

## Build

```
$ ./gradlew gem  # -t to watch change of files and rebuild continuously
```


CHANGELOG.md
Release 0.1.1 - 2020-03-11

* Built with the "org.embulk.embulk-plugins" Gradle plugin.
* Use its own SnakeYAML library.

Release 0.1.0 - 2018-08-08

* The first release.


src/test/java/org/embulk/filter/decrypt/MockPageOutputReader.java
package org.embulk.filter.decrypt;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectReader;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import org.embulk.spi.Column;
import org.embulk.spi.Page;
import org.embulk.spi.PageReader;
import org.embulk.spi.Schema;
import org.embulk.spi.TestPageBuilderReader;
import org.embulk.spi.type.Types;

import java.io.IOException;

public class MockPageOutputReader
{
    private static final ObjectReader jsonReader = new ObjectMapper().reader();

    private MockPageOutputReader()
    {
    }

    public static ArrayNode readPageOutput(Schema schema, TestPageBuilderReader.MockPageOutput pageOutput) throws IOException
    {
        ArrayNode nodes = (ArrayNode) jsonReader.createArrayNode();
        for (Page page : pageOutput.pages) {
            nodes.addAll(pageToJsonArray(schema, page));
        }
        return nodes;
    }

    private static ArrayNode pageToJsonArray(Schema schema, Page page) throws IOException
    {
        ArrayNode nodes = (ArrayNode) jsonReader.createArrayNode();
        PageReader reader = new PageReader(schema);
        reader.setPage(page);
        while (reader.nextRecord()) {
            ObjectNode node = (ObjectNode) jsonReader.createObjectNode();
            for (Column column : schema.getColumns()) {
                if (Types.STRING == column.getType()) {
                    if (!reader.isNull(column)) {
                        node.put(column.getName(), reader.getString(column));
                    }
                }
                else if (Types.BOOLEAN == column.getType()) {
                    node.put(column.getName(), reader.getBoolean(column));
                }
                else if (Types.DOUBLE == column.getType()) {
                    node.put(column.getName(), reader.getDouble(column));
                }
                else if (Types.JSON == column.getType()) {
                    if (!reader.isNull(column)) {
                        node.set(column.getName(), jsonReader.readTree(reader.getJson(column).toJson()));
                    }
                }
                else if (Types.LONG == column.getType()) {
                    node.put(column.getName(), reader.getLong(column));
                }
                else if (Types.TIMESTAMP == column.getType()) {
                    if (!reader.isNull(column)) {
                        node.put(column.getName(), reader.getTimestamp(column).toEpochMilli());
                    }
                }
            }
            nodes.add(node);
        }
        return nodes;
    }
}


src/test/java/org/embulk/filter/decrypt/TestDefaultRetryable.java
package org.embulk.filter.decrypt;

import org.embulk.EmbulkTestRuntime;
import org.embulk.spi.util.RetryExecutor;
import org.embulk.spi.util.RetryExecutor.RetryGiveupException;
import org.junit.Rule;
import org.junit.Test;

import java.io.IOException;
import java.util.concurrent.Callable;

import static java.lang.String.format;
import static org.msgpack.core.Preconditions.checkArgument;

public class TestDefaultRetryable
{
    @Rule
    public EmbulkTestRuntime runtime = new EmbulkTestRuntime();  // require for DefaultRetryable's logger

    private static class Deny extends RuntimeException implements Callable
    {
        private int pastCalls = 0;
        private final int targetCalls;
        private Exception exception;

        Deny(int targetCalls)
        {
            super(format("Try harder! (Will pass after %d calls)", targetCalls));
            checkArgument(targetCalls >= 0);
            this.targetCalls = targetCalls;
        }

        static Deny until(int calls)
        {
            return new Deny(calls);
        }

        Deny with(Exception exception)
        {
            this.exception = exception;
            return this;
        }

        @Override
        public Object call() throws Exception
        {
            if (pastCalls < targetCalls) {
                pastCalls++;
                if (exception != null) {
                    throw exception;
                }
                else {
                    throw this;
                }
            }
            pastCalls++;
            return null;
        }
    }

    private static RetryExecutor retryExecutor()
    {
        return RetryExecutor.retryExecutor()
                .withInitialRetryWait(0)
                .withMaxRetryWait(0);
    }

    @Test
    @SuppressWarnings("unchecked")
    public void guaranteeRetryAttemptsJustLikeRetryable() throws Exception
    {
        retryExecutor()
                .withRetryLimit(0)
                .run(new DefaultRetryable(Deny.until(0)));
        retryExecutor()
                .withRetryLimit(1)
                .run(new DefaultRetryable(Deny.until(1)));
        retryExecutor()
                .withRetryLimit(2)
                .run(new DefaultRetryable(Deny.until(1)));
        retryExecutor()
                .withRetryLimit(3)
                .run(new DefaultRetryable(Deny.until(2)));
    }

    @Test(expected = RetryGiveupException.class)
    @SuppressWarnings("unchecked")
    public void failAfterExceedingAttemptsJustLikeRetryable() throws Exception
    {
        retryExecutor()
                .withRetryLimit(3)
                .run(new DefaultRetryable(Deny.until(4)));
    }

    @Test(expected = Deny.class)
    @SuppressWarnings("unchecked")
    public void executeShouldUnwrapRetryGiveupException() throws Exception
    {
        new DefaultRetryable(Deny.until(4))
                .executeWith(retryExecutor().withRetryLimit(3));
    }

    @Test(expected = RuntimeException.class)
    @SuppressWarnings("unchecked")
    public void executeShouldUnwrapRetryGiveupExceptionButRewrapCheckedExceptionInARuntimeException()
    {
        new DefaultRetryable(Deny.until(4).with(new Exception("A checked exception")))
                .executeWith(retryExecutor().withRetryLimit(3));
    }

    @Test(expected = IOException.class)
    public void executeAndPropagateAsIsShouldLeaveOriginalExceptionUnwrapped() throws IOException
    {
        RetryExecutor retryExc = retryExecutor().withRetryLimit(3);
        // An explicit type parameter for operation return type is needed here,
        // Without one, javac (at least on 1.8) will fails to infer the X exception type parameter.
        new DefaultRetryable<Object>() {
            @Override
            public Object call() throws IOException
            {
                throw new IOException();
            }
        }.executeWithCheckedException(retryExc, IOException.class);
    }

    @Test(expected = IllegalStateException.class)
    public void executeWithoutAnImplementationShouldThrowAnIllegalStateException()
    {
        new DefaultRetryable().executeWith(retryExecutor());
    }
}


src/main/java/org/embulk/filter/decrypt/DefaultRetryable.java
package org.embulk.filter.decrypt;

import com.amazonaws.AmazonServiceException;
import com.google.common.base.Throwables;
import org.apache.http.HttpStatus;
import org.embulk.spi.Exec;
import org.embulk.spi.util.RetryExecutor;
import org.slf4j.Logger;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.Callable;

import static java.lang.String.format;
import static org.embulk.spi.util.RetryExecutor.RetryGiveupException;
import static org.embulk.spi.util.RetryExecutor.Retryable;

/**
 * Retryable utility, regardless the occurred exceptions,
 * Also provide a default approach for exception propagation.
 */
class DefaultRetryable<T> implements Retryable<T>
{
    private static final Logger log = Exec.getLogger(DefaultRetryable.class);
    private static final Set<Integer> NONRETRYABLE_STATUS_CODES = new HashSet<Integer>(2);
    private static final Set<String> NONRETRYABLE_ERROR_CODES = new HashSet<String>(1);
    private String operationName;
    private Callable<T> callable;

    static {
        NONRETRYABLE_STATUS_CODES.add(HttpStatus.SC_FORBIDDEN);
        NONRETRYABLE_STATUS_CODES.add(HttpStatus.SC_METHOD_NOT_ALLOWED);
        NONRETRYABLE_ERROR_CODES.add("ExpiredToken");
        NONRETRYABLE_ERROR_CODES.add("NoSuchBucket");
        NONRETRYABLE_ERROR_CODES.add("NoSuchKey");
    }

    /**
     * @param operationName the name that will be referred on logging
     */
    public DefaultRetryable(String operationName)
    {
        this.operationName = operationName;
    }

    /**
     * @param operationName the name that will be referred on logging
     * @param callable the operation, either define this at construction time or override the call() method
     */
    public DefaultRetryable(String operationName, Callable<T> callable)
    {
        this.operationName = operationName;
        this.callable = callable;
    }

    public DefaultRetryable()
    {
        this("Anonymous operation");
    }

    public DefaultRetryable(Callable<T> callable)
    {
        this("Anonymous operation", callable);
    }

    @Override
    public T call() throws Exception
    {
        if (callable != null) {
            return callable.call();
        }
        else {
            throw new IllegalStateException("Either override call() or construct with a Runnable");
        }
    }

    @Override
    public boolean isRetryableException(Exception exception)
    {
        // No retry on a subset of service exceptions
        if (exception instanceof AmazonServiceException) {
            AmazonServiceException ase = (AmazonServiceException) exception;
            return !NONRETRYABLE_STATUS_CODES.contains(ase.getStatusCode()) && !NONRETRYABLE_ERROR_CODES.contains(ase.getErrorCode());
        }
        return true;
    }

    @Override
    public void onRetry(Exception exception, int retryCount, int retryLimit, int retryWait)
    {
        String message = format("%s failed. Retrying %d/%d after %d seconds. Message: %s",
                operationName, retryCount, retryLimit, retryWait / 1000, exception.getMessage());
        if (retryCount % retryLimit == 0) {
            log.warn(message, exception);
        }
        else {
            log.warn(message);
        }
    }

    @Override
    public void onGiveup(Exception firstException, Exception lastException)
    {
        // Exceptions would be propagated, so it's up to the caller to handle, this is just warning
        log.warn("Giving up on retrying for {}, first exception is [{}], last exception is [{}]",
                operationName, firstException.getMessage(), lastException.getMessage());
    }

    /**
     * Run itself by the supplied executor,
     *
     * This propagates all exceptions (as unchecked) and unwrap RetryGiveupException for the original cause.
     * If the original exception already is a RuntimeException, it will be propagated as is. If not, it will
     * be wrapped around with a RuntimeException.
     *
     * For convenient, it execute normally without retrying when executor is null.
     *
     * @throws RuntimeException the original cause
     */
    public T executeWith(RetryExecutor executor)
    {
        if (executor == null) {
            try {
                return this.call();
            }
            catch (Exception e) {
                Throwables.propagate(e);
            }
        }

        try {
            return executor.runInterruptible(this);
        }
        catch (RetryGiveupException e) {
            throw Throwables.propagate(e.getCause());
        }
        catch (InterruptedException e) {
            throw Throwables.propagate(e);
        }
    }

    /**
     * Run itself by the supplied executor,
     *
     * Same as `executeWith`, this propagates all original exceptions. But `propagateAsIsException` will
     * be re-throw without being wrapped on a RuntimeException, whether it is a checked or unchecked exception.
     *
     * For convenient, it execute normally without retrying when executor is null.
     *
     * @throws X whatever checked exception that you decided to propagate directly
     * @throws RuntimeException wrap around whatever the original cause of failure (potentially thread interruption)
     */
    public <X extends Throwable> T executeWithCheckedException(RetryExecutor executor,
                                                               Class<X> propagateAsIsException) throws X
    {
        if (executor == null) {
            try {
                return this.call();
            }
            catch (Exception e) {
                Throwables.propagate(e);
            }
        }

        try {
            return executor.runInterruptible(this);
        }
        catch (RetryGiveupException e) {
            Throwables.propagateIfInstanceOf(e.getCause(), propagateAsIsException);
            throw Throwables.propagate(e.getCause());
        }
        catch (InterruptedException e) {
            throw Throwables.propagate(e);
        }
    }
}


src/main/java/org/embulk/filter/decrypt/DecryptFilterPlugin.java
package org.embulk.filter.decrypt;

import com.amazonaws.AmazonServiceException;
import com.amazonaws.SdkClientException;
import com.amazonaws.auth.AWSCredentials;
import com.amazonaws.auth.AWSCredentialsProvider;
import com.amazonaws.auth.BasicAWSCredentials;
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.GetObjectRequest;
import com.amazonaws.services.s3.model.S3Object;
import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.google.common.annotations.VisibleForTesting;
import com.google.common.base.Optional;
import com.google.common.io.BaseEncoding;
import org.embulk.config.Config;
import org.embulk.config.ConfigDefault;
import org.embulk.config.ConfigException;
import org.embulk.config.ConfigSource;
import org.embulk.config.Task;
import org.embulk.config.TaskSource;
import org.embulk.spi.Column;
import org.embulk.spi.ColumnVisitor;
import org.embulk.spi.DataException;
import org.embulk.spi.Exec;
import org.embulk.spi.FilterPlugin;
import org.embulk.spi.Page;
import org.embulk.spi.PageBuilder;
import org.embulk.spi.PageOutput;
import org.embulk.spi.PageReader;
import org.embulk.spi.Schema;
import org.embulk.util.snakeyaml.EmbulkYamlTagResolver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.yaml.snakeyaml.DumperOptions;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.SafeConstructor;
import org.yaml.snakeyaml.representer.Representer;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import java.io.IOException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.EnumSet;
import java.util.List;
import java.util.Map;

import static com.google.common.base.Charsets.UTF_8;
import static com.google.common.base.Strings.isNullOrEmpty;
import static java.lang.String.format;
import static org.apache.commons.lang3.StringUtils.join;

public class DecryptFilterPlugin
        implements FilterPlugin
{
    public enum Algorithm
    {
        AES_256_CBC("AES/CBC/PKCS5Padding", "AES", 256, true, "AES", "AES-256", "AES-256-CBC"),
        AES_192_CBC("AES/CBC/PKCS5Padding", "AES", 192, true, "AES-192", "AES-192-CBC"),
        AES_128_CBC("AES/CBC/PKCS5Padding", "AES", 128, true, "AES-128", "AES-128-CBC"),
        AES_256_ECB("AES/ECB/PKCS5Padding", "AES", 256, false, "AES-256-ECB"),
        AES_192_ECB("AES/ECB/PKCS5Padding", "AES", 192, false, "AES-192-ECB"),
        AES_128_ECB("AES/ECB/PKCS5Padding", "AES", 128, false, "AES-128-ECB");

        private final String javaName;
        private final String javaKeySpecName;
        private final int keyLength;
        private final boolean useIv;
        private String[] displayNames;

        Algorithm(String javaName, String javaKeySpecName, int keyLength, boolean useIv, String... displayNames)
        {
            this.javaName = javaName;
            this.javaKeySpecName = javaKeySpecName;
            this.keyLength = keyLength;
            this.useIv = useIv;
            this.displayNames = displayNames;
        }

        public String getJavaName()
        {
            return javaName;
        }

        public String getJavaKeySpecName()
        {
            return javaKeySpecName;
        }

        public int getKeyLength()
        {
            return keyLength;
        }

        public boolean useIv()
        {
            return useIv;
        }

        @JsonCreator
        public static Algorithm fromName(String name)
        {
            EnumSet<Algorithm> algos = EnumSet.allOf(Algorithm.class);
            for (Algorithm algo : algos) {
                for (String n : algo.displayNames) {
                    if (n.equals(name)) {
                        return algo;
                    }
                }
            }
            throw new ConfigException(format("Unsupported algorithm '%s'. Supported algorithms are %s",
                    name,
                    join(algos, ", ")));
        }

        @JsonValue
        @Override
        public String toString()
        {
            return displayNames[displayNames.length - 1];
        }
    }

    public enum Encoder
    {
        BASE64("base64", BaseEncoding.base64()),
        HEX("hex", BaseEncoding.base16());

        private final BaseEncoding encoding;
        private final String name;

        Encoder(String name, BaseEncoding encoding)
        {
            this.name = name;
            this.encoding = encoding;
        }

        public byte[] decode(String s)
        {
            return encoding.decode(s);
        }

        @JsonCreator
        public static Encoder fromName(String name)
        {
            EnumSet<Encoder> encoders = EnumSet.allOf(Encoder.class);
            for (Encoder encoder : encoders) {
                if (encoder.name.equals(name)) {
                    return encoder;
                }
            }
            throw new ConfigException(
                    format("Unsupported input encoding '%s'. Supported encodings are %s.",
                            name,
                            join(encoders, ", ")));
        }

        @JsonValue
        @Override
        public String toString()
        {
            return name;
        }
    }

    public enum KeyType
    {
        INLINE,
        S3;

        @JsonCreator
        public static KeyType of(String value)
        {
            return KeyType.valueOf(value.toUpperCase());
        }

        @Override
        @JsonValue
        public String toString()
        {
            return super.toString().toLowerCase();
        }
    }

    public interface PluginTask
            extends Task
    {
        @Config("algorithm")
        public Algorithm getAlgorithm();

        @Config("input_encoding")
        @ConfigDefault("\"base64\"")
        public Encoder getInputEncoding();

        @Config("key_type")
        @ConfigDefault("\"inline\"")
        KeyType getKeyType();

        @Config("key_hex")
        @ConfigDefault("null")
        public Optional<String> getKeyHex();

        public void setKeyHex(Optional<String> key);

        @Config("iv_hex")
        @ConfigDefault("null")
        public Optional<String> getIvHex();

        public void setIvHex(Optional<String> iv);

        @Config("aws_params")
        @ConfigDefault("null")
        public Optional<AWSParams> getAWSParams();

        @Config("column_names")
        public List<String> getColumnNames();
    }

    public interface AWSParams extends Task
    {
        @Config("region")
        public String getRegion();

        @Config("access_key")
        public String getAccessKey();

        @Config("secret_key")
        public String getSecretKey();

        @Config("bucket")
        public String getBucket();

        @Config("path")
        public String getPath();
    }

    private static final Yaml yaml = new Yaml(new SafeConstructor(), new Representer(), new DumperOptions(), new EmbulkYamlTagResolver());
    private static final Logger log = LoggerFactory.getLogger(DecryptFilterPlugin.class);

    @Override
    public void transaction(ConfigSource config, Schema inputSchema,
                            FilterPlugin.Control control)
    {
        PluginTask task = config.loadConfig(PluginTask.class);

        validateAndResolveKey(task, inputSchema);

        control.run(task.dump(), inputSchema);
    }

    @VisibleForTesting
    public Map<String, String> retrieveKey(final String bucket, final String path, final AmazonS3 client)
    {
        S3Object fullObject = null;

        try {
            fullObject = client.getObject(new GetObjectRequest(bucket, path));
            if (fullObject == null) {
                throw new ConfigException("S3 key file is not enabled to be retrieved");
            }
            return (Map<String, String>) yaml.load(fullObject.getObjectContent());
        }
        catch (AmazonServiceException e) {
            // The call was transmitted successfully, but Amazon S3 couldn't process
            // it, so it returned an error response.
            if (e.getErrorType().equals(AmazonServiceException.ErrorType.Client)) {
                // HTTP 40x errors. auth error, bucket doesn't exist, etc. See AWS document for the full list:
                // http://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html
                if (e.getStatusCode() != 400
                        || "ExpiredToken".equalsIgnoreCase(e.getErrorCode())) {
                    throw new ConfigException(e);
                }
            }
            throw e;
        }
        catch (ClassCastException e) {
            throw new ConfigException("S3 key file content is unexpected format");
        }
        finally {
            // To ensure that the network connection doesn't remain open, close any open input streams.
            if (fullObject != null) {
                try {
                    fullObject.close();
                }
                catch (IOException e) {
                    log.warn("Failure to close S3 Object input stream", e);
                }
            }
        }
    }

    @VisibleForTesting
    public AmazonS3 newS3Client(final AWSParams awsParams)
    {
        AWSCredentialsProvider awsCredentialsProvider = new AWSCredentialsProvider()
        {
            @Override
            public AWSCredentials getCredentials()
            {
                return new BasicAWSCredentials(awsParams.getAccessKey(), awsParams.getSecretKey());
            }

            @Override
            public void refresh()
            {
            }
        };

        try {
            return AmazonS3ClientBuilder.standard()
                    .withRegion(awsParams.getRegion())
                    .withCredentials(awsCredentialsProvider)
                    .build();
        }
        catch (SdkClientException e) {
            throw new ConfigException(e);
        }
    }

    @Override
    public PageOutput open(TaskSource taskSource, final Schema inputSchema,
                           final Schema outputSchema, final PageOutput output)
    {
        final PluginTask task = taskSource.loadTask(PluginTask.class);

        final Cipher cipher;
        try {
            cipher = getCipher(Cipher.DECRYPT_MODE, task);
        }
        catch (Exception e) {
            throw new DataException(e);
        }

        final int[] targetColumns = new int[task.getColumnNames().size()];
        int i = 0;
        for (String name : task.getColumnNames()) {
            targetColumns[i++] = inputSchema.lookupColumn(name).getIndex();
        }

        return new PageOutput() {
            private final PageReader pageReader = new PageReader(inputSchema);
            private final PageBuilder pageBuilder = new PageBuilder(Exec.getBufferAllocator(), outputSchema, output);
            private final Encoder encoder = task.getInputEncoding();

            @Override
            public void finish()
            {
                pageBuilder.finish();
            }

            @Override
            public void close()
            {
                pageBuilder.close();
            }

            private boolean isTargetColumn(Column c)
            {
                for (int i = 0; i < targetColumns.length; i++) {
                    if (c.getIndex() == targetColumns[i]) {
                        return true;
                    }
                }
                return false;
            }

            @Override
            public void add(Page page)
            {
                pageReader.setPage(page);

                while (pageReader.nextRecord()) {
                    inputSchema.visitColumns(new ColumnVisitor() {
                        @Override
                        public void booleanColumn(Column column)
                        {
                            if (pageReader.isNull(column)) {
                                pageBuilder.setNull(column);
                            }
                            else {
                                pageBuilder.setBoolean(column, pageReader.getBoolean(column));
                            }
                        }

                        @Override
                        public void longColumn(Column column)
                        {
                            if (pageReader.isNull(column)) {
                                pageBuilder.setNull(column);
                            }
                            else {
                                pageBuilder.setLong(column, pageReader.getLong(column));
                            }
                        }

                        @Override
                        public void doubleColumn(Column column)
                        {
                            if (pageReader.isNull(column)) {
                                pageBuilder.setNull(column);
                            }
                            else {
                                pageBuilder.setDouble(column, pageReader.getDouble(column));
                            }
                        }

                        @Override
                        public void stringColumn(Column column)
                        {
                            if (pageReader.isNull(column)) {
                                pageBuilder.setNull(column);
                            }
                            else if (isTargetColumn(column)) {
                                String orig = pageReader.getString(column);
                                byte[] decoded = encoder.decode(orig);
                                byte[] decrypted;

                                try {
                                    decrypted = cipher.doFinal(decoded);
                                }
                                catch (BadPaddingException ex) {
                                    // this must not happen because PKCS5Padding is always enabled
                                    throw new DataException(ex);
                                }
                                catch (IllegalBlockSizeException ex) {
                                    // this must not happen because always doFinal is called
                                    throw new DataException(ex);
                                }
                                pageBuilder.setString(column, new String(decrypted, UTF_8));
                            }
                            else {
                                pageBuilder.setString(column, pageReader.getString(column));
                            }
                        }

                        @Override
                        public void timestampColumn(Column column)
                        {
                            if (pageReader.isNull(column)) {
                                pageBuilder.setNull(column);
                            }
                            else {
                                pageBuilder.setTimestamp(column, pageReader.getTimestamp(column));
                            }
                        }

                        @Override
                        public void jsonColumn(Column column)
                        {
                            if (pageReader.isNull(column)) {
                                pageBuilder.setNull(column);
                            }
                            else {
                                pageBuilder.setJson(column, pageReader.getJson(column));
                            }
                        }
                    });
                    pageBuilder.addRecord();
                }
            }
        };
    }

    private void validateAndResolveKey(PluginTask task, Schema schema) throws ConfigException
    {
        switch (task.getKeyType()) {
            case INLINE:
                if (!task.getKeyHex().isPresent()) {
                    throw new ConfigException("Field 'key_hex' is required but not set");
                }
                if (task.getAlgorithm().useIv() && !task.getIvHex().isPresent()) {
                    throw new ConfigException("Algorithm '" + task.getAlgorithm() + "' requires initialization vector. Please generate one and set it to iv_hex option.");
                }
                else if (!task.getAlgorithm().useIv() && task.getIvHex().isPresent()) {
                    log.warn("Algorithm '" + task.getAlgorithm() + "' doesn't use initialization vector. iv_hex is ignored");
                }
                break;
            case S3:
                if (!task.getAWSParams().isPresent()) {
                    throw new ConfigException("AWS Params are required for S3 Key type");
                }
                AWSParams params = task.getAWSParams().get();
                AmazonS3 s3Client = newS3Client(params);
                Map<String, String> keys = retrieveKey(params.getBucket(), params.getPath(), s3Client);
                String key = keys.get("key_hex");
                if (isNullOrEmpty(key)) {
                    throw new ConfigException("Field 'key_hex' is required but not set");
                }
                String iv = keys.get("iv_hex");
                if (task.getAlgorithm().useIv() && isNullOrEmpty(iv)) {
                    throw new ConfigException("Algorithm '" + task.getAlgorithm() + "' requires initialization vector. Please generate one and set it to iv_hex option.");
                }
                else if (!task.getAlgorithm().useIv() && !isNullOrEmpty(iv)) {
                    log.warn("Algorithm '" + task.getAlgorithm() + "' doesn't use initialization vector. iv_hex is ignored");
                }
                task.setKeyHex(Optional.of(key));
                if (!isNullOrEmpty(iv)) {
                    task.setIvHex(Optional.of(iv));
                }
                break;
            default:
                throw new ConfigException(String.format("Key type [%s] is not supported", task.getKeyType().toString()));
        }

        // Validate Cipher
        try {
            getCipher(Cipher.DECRYPT_MODE, task);
        }
        catch (Exception e) {
            throw new ConfigException(e);
        }

        // validate column_names
        for (String name : task.getColumnNames()) {
            schema.lookupColumn(name);
        }
    }

    private Cipher getCipher(int mode, PluginTask task)
            throws NoSuchAlgorithmException, InvalidAlgorithmParameterException, NoSuchPaddingException, InvalidKeyException
    {
        Algorithm algo = task.getAlgorithm();

        byte[] keyData = BaseEncoding.base16().decode(task.getKeyHex().get());
        SecretKeySpec key = new SecretKeySpec(keyData, algo.getJavaKeySpecName());

        if (algo.useIv()) {
            byte[] ivData = BaseEncoding.base16().decode(task.getIvHex().get());
            IvParameterSpec iv = new IvParameterSpec(ivData);

            Cipher cipher = Cipher.getInstance(algo.getJavaName());
            cipher.init(mode, key, iv);
            return cipher;
        }
        else {
            Cipher cipher = Cipher.getInstance(algo.getJavaName());
            cipher.init(mode, key);
            return cipher;
        }
    }
}
